src/mon/NVMeofGwMap.cc

```diff
@@ -0,0 +1,600 @@
+                start_timer(gw_state.first, group_key, grpid, 30);
+                st.blocklist_data[grpid].osd_epoch = epoch;
+                st.blocklist_data[grpid].is_failover = false;
+                dout(10)  << "Expired Failback-preparation timer from GW " << gw_id << " ANA groupId "<< grpid << "osd epoch "  << st.blocklist_data[grpid].osd_epoch  << dendl;
+
+                dout(10) << "Starting blocklist: Failback from GW " << gw_id << " to " << gw_state.first << dendl;
+                map_modified = true;
+                break;
+            }
+            else if (st.ana_grp_id == grpid){
+                if(st.sm_state[grpid] == GW_STATES_PER_AGROUP_E::GW_STANDBY_STATE  &&  st.availability == GW_AVAILABILITY_E::GW_AVAILABLE) {
+                    st.active_state(grpid);// GW failed and started during the persistency interval
+                    dout(4)  << "Failback unsuccessfull. GW: " << gw_state.first << "becomes Active for the ana group " << grpid  << dendl;
+                }
+                fbp_gw_state.standby_state(grpid);
+                dout(4) << "Failback unsuccessfull GW: " << gw_id << "becomes standby for the ana group " << grpid  << dendl;
+                map_modified = true;
+                break;
+            }
+        }
+    }
+    else if(fbp_gw_state.sm_state[grpid] == GW_STATES_PER_AGROUP_E::GW_WAIT_BLOCKLIST_CMPL){
+        dout(1) << " Expired GW_WAIT_FAILOVER_PREPARED timer from GW " << gw_id << " ANA groupId: "<< grpid << dendl;
+        ceph_assert(false);
+    }
+}
+
+NvmeGwCreated& NVMeofGwMap::find_already_created_gw(const NvmeGwId &gw_id, const NvmeGroupKey& group_key)
+{
+    auto& group_gws = Created_gws[group_key];
+    auto  gw_it = group_gws.find(gw_id);
+    ceph_assert(gw_it != group_gws.end());//should not happen
+    return gw_it->second;
+}
+
+struct CMonRequestProposal : public Context {
+  NVMeofGwMap *m;
+  CMonRequestProposal(NVMeofGwMap *mon) : m(mon) {}
+  void finish(int r) {
+      dout(4) << "osdmon is  writable? " << m->mon->osdmon()->is_writeable() << dendl;
+      if(m->mon->osdmon()->is_writeable()){
+        m->mon->nvmegwmon()->request_proposal(m->mon->osdmon());
+      }
+      else {
+          m->mon->osdmon()->wait_for_writeable_ctx( new CMonRequestProposal(m));
+      }
+  }
+};
+
+int NVMeofGwMap::blocklist_gw(const NvmeGwId &gw_id, const NvmeGroupKey& group_key, NvmeAnaGrpId grpid, epoch_t &epoch, bool failover)
+{
+    NvmeGwCreated& gw_map =  Created_gws[group_key][gw_id];  //find_already_created_gw(gw_id, group_key);
+
+     if (gw_map.nonce_map[grpid].size() > 0){
+        NvmeNonceVector &nonce_vector = gw_map.nonce_map[grpid];;
+        if(failover == false)
+           nonce_vector = gw_map.copied_nonce_map[grpid];
+
+        std::string str = "[";
+        entity_addrvec_t addr_vect;
+        //auto until = ceph_clock_now();  // until += g_conf().get_val<double>("mon_mgr_blocklist_interval"); // 1day - TODO
+        utime_t until(30, 0);
```

Comment:
We can make the blocklist last forever. In any case, we recycle the cluster context relevant to these namespaces, when we want this GW to be able to write again to these namespaces. I think you explain to us that this timeout only defines how long the list will be published as a part of the map. But in practice, a cluster context is useless after it was block listed.

Reply:
> We can make the blocklist last forever.

Forever is the opposite extreme which we don't want because blocklist entries would pile up in the osdmap.

> In any case, we recycle the cluster context relevant to these namespaces, when we want this GW to be able to write again to these namespaces.

Can you point at where that is done?

> I think you explain to us that this timeout only defines how long the list will be published as a part of the map.

Right.

> But in practice, a cluster context is useless after it was block listed.

This goes only for new ops and also highly depends on which RBD features are enabled and other things.  For example, disabling exclusive lock on the image would make such a cluster context a lot less useless/more dangerous ;)

And regardless, anything pending in the lower layers is fair game...

Reply:
The recycling is done in the nvmeof GW code not in here. https://github.com/ceph/ceph-nvmeof/pull/445. 
We can increase the blocklist from 30 seconds to minutes if that will address your concern. The blocklist should make sure that after that, no new IO can reach Ceph using that cluster context. Before doing the blocklist, we also set the path to be inaccessible, so: 
1. SPDK will not accept new writes (and also the initiator will not try to send new write). 
2. We blocklist so ceph will not get new write.

Reply:
> The recycling is done in the nvmeof GW code not in here. [ceph/ceph-nvmeof#445](https://github.com/ceph/ceph-nvmeof/pull/445).

When `create_bdev()` is called after deleting the old bdev, is the cluster context that it gets from `_get_cluster()` guaranteed to be allocated at that point or can a previously allocated context be used?  I tried to follow the code, but it comes down to a dictionary keyed on a pair of `anagrp` and `cluster_name` and I'm missing that context.

> We can increase the blocklist from 30 seconds to minutes if that will address your concern.

Let's make it at least 1 hour (`mon_osd_blocklist_default_expire` default).

> The blocklist should make sure that after that, no new IO can reach Ceph using that cluster context. Before doing the blocklist, we also set the path to be inaccessible, so:
> 
> 1. SPDK will not accept new writes (and also the initiator will not try to send new write).
> 2. We blocklist so ceph will not get new write.

The trickiest bit is ensuring that the osdmap with the blocklist propagates.  Until that happens OSDs can very much get new writes and would happily apply them.

Reply:
@idryomov,   During Failback  scenario the below sequence of actions is applied :
1. NVMeofGWMonitor makes change in map and sets  state  GW_WAIT_FAILBACK_PREPARED to the GW that "returns the Ana group".  This state is propagated to the GW . "current GW ana group entity vector" is saved ! 

2. GW upon receiving the command does 2 actions : 
    a)  sets Inaccessible state to all namespaces related to "returned" ana group. This SPDK  action also immediately blocks all new IOs from the host
    b) recycles all namespaces related to the "returned" ana group. New cluster context(s)are created during this process.
     
3.  5 seconds later NVMeofGWMonitor   starts blocklist  for saved in step 1 "old entity vector". It starts waiting for the new osd epoch. 
4.  When osd epoch changes (wait for new epoch  is done on the client's side) the GW "ana group owner" of Ana group, receives command to set all namespaces on this group to Optimized state.  GW "ana group owner" takes over the ana group.

So this behavior should  guarantee the absence of writes from 2 GWs.

Reply:
> @idryomov, During Failback scenario the below sequence of actions is applied :

I think the more interesting case is failover (which would typically be unorderly/forced).  Failback is typically orderly, so it should be easier to deal with.

> 
> 1. NVMeofGWMonitor makes change in map and sets  state  GW_WAIT_FAILBACK_PREPARED to the GW that "returns the Ana group".  This state is propagated to the GW . "current GW ana group entity vector" is saved !
> 
> 2. GW upon receiving the command does 2 actions :
>        a)  sets Inaccessible state to all namespaces related to "returned" ana group. This SPDK  action also immediately blocks all new IOs from the host
>        b) recycles all namespaces related to the "returned" ana group. New cluster context(s)are created during this process.
> 
> 3. 5 seconds later NVMeofGWMonitor   starts blocklist  for saved in step 1 "old entity vector". It starts waiting for the new osd epoch.

So new cluster contexts (2b above) are created _before_ blocklist is even issued?

> 
> 4. When osd epoch changes (wait for new epoch  is done on the client's side) the GW "ana group owner", receives command to set all namespaces on this group to Optimized state.  GW "ana group owner" takes over the ana group.

Are you referring to the wait in `NVMeofGwMonitorClient::handle_nvmeof_gw_map()`?  If so, I don't think it affects the gateway at all.  I posted a related question on it yesterday, see https://github.com/ceph/ceph/pull/54671#discussion_r1521920217.

Reply:
> So new cluster contexts (2b above) are created before the blocklist is even issued?
Yes, but the monitor save the "old" contexts, and the blocklist is to the old contexts - not to the new ones.

Reply:
> > So new cluster contexts (2b above) are created before the blocklist is even issued?
> 
> Yes, but the monitor save the "old" contexts, and the blocklist is to the old contexts - not to the new ones.

This is where
> The trickiest bit is ensuring that the osdmap with the blocklist propagates.

concern from my earlier comment comes in.  If the new cluster context is created before the osdmap with the blocklist is seen, there is no guarantee that the blocklist would be enforced by the OSD.

Reply:
we don't allow any client to be write to the namespaces before we get the new osd map epoch (epoch that is > the epoch before the blocklist). Is there any other way to verify that the client knows about the new OSD map?

Reply:
> we don't allow any client to be write to the namespaces before we get the new osd map epoch (epoch that is > the epoch before the blocklist).

Again, what is "we" -- `NVMeofGwMonitorClient` or the actual gateway?  This is the question I asked Leonid above.

Reply:
The gateway client wait for the new epoch, and notifies the monitor that new epoch was received on its side. Then the monitor responds and allows that client to open the access to these namespaces (i.e. set to optimized). @leonidc please correct me if I'm wrong.

Reply:
NVMeofGwMonitorClient  just forwards the   decisions of NVMeofGWMon and is responsible for waiting for the new osd map epoch .

Reply:
@leonidc I need you to answer my outstanding question first:
> Are you referring to the wait in NVMeofGwMonitorClient::handle_nvmeof_gw_map()?

Is there any other wait for the osdmap in the gateway or that piece in `NVMeofGwMonitorClient` is all there is?

Reply:
Failover/failback processes per each ana-group are handled by separate state machines in the NVMeofGWMon. When state per some ana-group delivered to the GW-client is GW_WAIT_BLOCKLIST_CMPL,  Client in **handle_nvmeof_gw_map**  starts to wait for epoch changed, it takes ~3 cycles = 6seconds. 
The code that waits  : bool const ready = objecter.with_osdmap( ...

Reply:
> The code that waits : bool const ready = objecter.with_osdmap( ...

OK, I'm going to interpret that as "No, there are no other waits for a new enough osdmap in the gateway, the one in `NVMeofGwMonitorClient` is all there is".  Based on that and Aviv confirming that new cluster contexts are created before the blocklist is even issued in https://github.com/ceph/ceph/pull/54671#discussion_r1522913809, it seems like the following can happen:

0. Assume that the current osdmap has epoch 10 (e10) and everything else in the system has it
1. Assume that there is an in-flight write (W1) tagged with e10 somewhere in librbd or the messenger under cluster context 1 (CC1)
2. A new cluster context (CC2) is created as part of recycling, it also gets e10
3. Blocklist for CC1 is proposed and gets published in e11
4. The wait in `NVMeofGwMonitorClient::handle_nvmeof_gw_map()` kicks in, `NVMeofGwMonitorClient` gets e11 but CC1, CC2 and the OSD for which W1 is targeted are all still on e10
5. After seeing e11, `NVMeofGwMonitorClient` allows a namespace/bdev that would use CC2 to go live (transition to OPTIMIZED state)
6. This namespace/bdev accepts a new write (W2) which overlaps with W1 (i.e. covers the same block range as W1 and targets the same OSD)
7. When W2 is received by librbd, it gets tagged with e10 because CC2 is still on e10, not e11
8. Some time passes, W2 reaches the OSD and gets applied -- W2 is tagged with e10 and the OSD still has only on e10, so everything is in order from the OSD perspective
9. W2 is completed and acked to the user
10. More time passes, W1 finally reaches the OSD and also gets applied despite the blocklist in e11 -- W1 is tagged with e10 and the OSD still has only e10, so everything is in order from the OSD perspective again!  And voila, a pre-blocklist write has trashed a post-blocklist write which was acked to the user.

(I have skipped the ANA group stuff and focused on the interaction with Ceph -- this is just to seed the discussion.)

Reply:
I don't think that this scenario is accurate the way its described here relative to the design and code. Maybe we need to do a discussion to explain.

Reply:
@caroav If the real nvme gateway doesn't rely on the OSDMap the "NVMeGwMonitorClient" has, then the MonitorClient's OSDMap epoch has no impact on the IOs that the nvme gateway is performing. This is one of a host of likely synchronization problems that are going to turn up by having them as separate daemons. :/

Reply:
Forget about fallback. What happens when you have GW1 and GW2, and GW2 times out in some way? You can't tell GW2 to stop doing IO because you can't communicate with it (but that *does not* mean that GW2 is not continuing to perform IOs!).

Reply:
> in any case, the code is flexible, we can make changes if you still believe that there is an issue. But it seems to be really and edge case that should not block all of this code from being delivered.

If it's an edge case that can lead to data corruption, it should absolutely block everything IMO.  Isn't the entire point of this PR (leveraging a Ceph monitor, introducing a new PAXOS service, etc) to avoid such edge cases?

Reply:
> OK so this is a failover scenario. What we do in failover is as follows:
> 
> 1. The monitor **thinks** that GW2 is down ***.
> 2. The monitor will blocklist GW1 cluster contexts (we have this ready as we get this in beacons).

The monitor will blocklist GW1 but not GW2 at some point? Why would you blocklist the *up* GW at all? Or is this a typo?

> 3. The monitor will send the GWs the map, which will indicate to that GW1 needs to switch to optimized on these namespaces.
> 4. The GW (i.e. monitor client) will wait to see that the new epoch (post blocklist) is arriving, and move to optimized.
> 
> *** For the case that GW2 is not really down. It is getting KA constantly from the monitor, and in case it is missing few KA, it should commit suicide. To avoid this split brain case. 

What is KA? You mean some kind of heartbeat?
Does correct behavior depend on the down GW2 killing itself?

> We assume that if the GW cannot get messages from Ceph, it also cannot get messages from nvmeof monitor.
What is "Ceph" in this case? Which components are these statements about? We have absolutely seen incidents where nodes lost communication to the monitor but were able to continue doing IO to OSDs.

Reply:
the monitor performs the blocklist on **GW2**  ana-grps cluster's contexts , there maybe several ana groups in optimized state at GW2 before it fails

Reply:
> The monitor will blocklist GW1 but not GW2 at some point? Why would you blocklist the up GW at all? Or is this a typo?

Typo 

> What is "Ceph" in this case? Which components are these statements about? We have absolutely seen incidents where nodes lost communication to the monitor but were able to continue doing IO to OSDs.

The monitor client is the part that send beacons to the monitor and gets the heartbeat. In case it misses few heartbeat it will cause the gw to die. 

> Does correct behavior depend on the down GW2 killing itself? 

In the case that the GW is really dead, then no issue. In the case that GW cannot be communicated from the monitor, and cannot probably also not getting the updated OSD maps, yes. What else could be done in such case? If we had some way to fence it physically - great, but I guess we don't?

Reply:
Okay, so what I understand your story to be from https://github.com/ceph/ceph/pull/54671#discussion_r1523505397, and my knowledge of the Ceph systems, is the following
1) GW2 times out for whatever reason
2) The monitor issues a new OSDMap blocklisting GW1 and sends a new NVMeGWMap to the NVMeGwMonitorClients with live sessions (I have not checked the code for if this actually works yet, but I presume you have them subscribing to map updates?)
3) The NVMeGwMonitorClient-1 waits for an osdmap containing the blocklist
4) NVMeGwMonitorClient-1 tells GW1 to start serving IO for the images previously owned by GW2

Unfortunately, what actually happens in the scenario above is something like the following:
0) The cluster and all clients are operating with osdmap 10
_1) GW2 times out for whatever reason_
_2) The monitor issues a new OSDMap **11** blocklisting GW1 and sends a new NVMeGWMap to the NVMeGwMonitorClients with live sessions (I have not checked the code for if this actually works yet, but I presume you have them subscribing to map updates?)_
_3) The NVMeGwMonitorClient-1 waits for **osdmap 11**_
_4) NVMeGwMonitorClient-1 tells GW1 to start serving IO for the images previously owned by GW2_
5) GW1 starts serving IO with osdmap 10, because it hasn't needed to get a newer map yet

And the possibility exists of
5a) GW2's host/router/whatever regains network access/unfreezes/something else, and an IO with osdmap 10 that was blocked is received and acted on by an OSD with map 10.
6) GW1 finally gets OSDMap 11

I tried searching for "osdmap" in the ceph-nvmeof repo to see if this is handled somehow and came across https://github.com/ceph/ceph-nvmeof/issues/193, which has a long discussion about how blocklisting works, but it doesn't seem to have been resolved or acted upon in a way that is obvious from searching.

Reply:
> The monitor issues a new OSDMap blocklisting GW1

gw2 not gw1 

> The monitor issues a new OSDMap 11 blocklisting GW1

gw2 not gw1 

With these corrections here, what is the issue that you see?

Reply:
> > The monitor issues a new OSDMap blocklisting GW1
> 
> gw2 not gw1
> 
> > The monitor issues a new OSDMap 11 blocklisting GW1
> 
> gw2 not gw1

Yep, now I'm the one typo-ing.

> With these corrections here, what is the issue that you see? Of course the GWs get map updates. The missing part (which we will fix as a bug that we need to add the commit suicide part. Its know to use regardless of this dicussion, but from time reasons we prefered to do it post merge. The heartbeats will be currently based on the acks that the nvmeof monitor sends to beacons, but we will do a separate mechanism of KA probably post this release.
> 
> As to 5a. this was supposed to be fixed by the self fencing of the GW,

I mean properly-applied blocklisting will resolve 5a, but you definitely can't rely on suicide to fix it. It is provably impossible to rely on that (though you should still implement it for a better UX and less likelihood of issues arising if there are other bugs).

Reply:
Ok but how can we properly apply blocklist in this case? I guess that no matter what will be the design, if you cannot make sure that the node is getting the messages, and you cannot physically kill it (fence it's electricity for example), what can you do?

Reply:
> > Does correct behavior depend on the down GW2 killing itself?
> 
> In the case that the GW is really dead, then no issue. In the case that GW cannot be communicated from the monitor, and cannot probably also not getting the updated OSD maps, yes. What else could be done in such case? If we had some way to fence it physically - great, but I guess we don't?

So properly-implemented blocklisting should resolve the problem. The issue here is really the disconnect with the real GW and the NVMeGwMonitorClient being able to have different map versions, and what appears to be the mistaken impression that a blocklist showing up in the OSDMap is enforced cluster-side before any client sees it.

Reply:
> Ok but how can we properly apply blocklist in this case? I guess that no matter what will be the design, if you cannot make sure that the node is getting the messages, and you cannot physically kill it (fence it's electricity for example), what can you do?

The real GW1 (not the NVMeGwMonitorClient) needs to get an OSDMap containing the blocklist, and use that osdmap for all communications which could conflict with GW2. Once you do that, the guarantee you get is that any object accessed (for read or write) by GW1 will know to blocklist IOs from GW2, and so you will not get conflicts visible to the client. (It is still possible that GW1 serves an IO at offset 20MB, and GW2 later completes a write at 400MB, but as the write was initiated we are allowed to complete it — we just can't let it complete out-of-order with another write to the same offset[s].)

Reply:
@gregsfortytwo how about the following fix:
When the NVMeGwMonitorClient get a new OSDMap it will call the gateway to refresh its OSDMap?
It is a small fix and we can add it in anther PR.
I want to merge this before the make check start failing again :(

Reply:
> The real GW1 (not the NVMeGwMonitorClient) needs to get an OSDMap containing the blocklist

So it means that we need to ensure that we see the required post blocklist epoch in the python process that actually generated the cluster contexts? Or any other way?

Reply:
> > The real GW1 (not the NVMeGwMonitorClient) needs to get an OSDMap containing the blocklist
> 
> So it means that we need to ensure that we see the required post blocklist epoch in the python process that actually generated the cluster contexts? Or any other way?

Yeah, the python process needs to see the appropriate map.

Reply:
Not in the python process but in the SPDK process, which is where these cluster contexts are created in response to `bdev_rbd_register_cluster` RPCs.  And this would need to be done for each cluster context individually, unless a cluster context is created after the blocklist osdmap is published (in this case the blocklist osdmap would be picked up in the process of creating a cluster context).

Reply:
These are...different processes?

Having this rely on a different process and code base that isn't in the PR makes life difficult. I'll have to go look at the documentation again and see if it's clear there.

Reply:
> These are...different processes?

Not only different processes but also different containers, I think.

Reply:
One container,  3 processes :
NVMe-of mon client (C++)
control gateway (python)
SPDk (c)

gRPC between the client and the control gateway
RPC between the gateway and SPDK